value-processor/src/org/immutables/value/processor/meta/ValueType.java

```diff
@@ -428,23 +428,34 @@ public CaseStructure getCases() {
     return includedElements;
   }
 
-  public List<CharSequence> passedAnnotations() {
-    return Annotations.getAnnotationLines(
-        element,
-        Sets.union(
-            constitution.style().passAnnotationsNames(),
-            constitution.style().additionalJsonAnnotationsNames()),
-        false,
-        ElementType.TYPE,
-        newTypeStringResolver(),
-        null);
+  @Nullable
+  private List<String> customAnnotations;
+
+  public List<String> passedAnnotations() {
+    if (customAnnotations == null) {
+      customAnnotations = Annotations.getAnnotationLines(
+          element,
+          Sets.union(
+              constitution.style().passAnnotationsNames(),
+              constitution.style().additionalJsonAnnotationsNames()),
+          false,
+          ElementType.TYPE,
+          newTypeStringResolver(),
+          null);
+    }
+
+    return customAnnotations;
   }
 
   private ImportsTypeStringResolver newTypeStringResolver() {
     @Nullable DeclaringType type = constitution.protoclass().declaringType().orNull();
     return new ImportsTypeStringResolver(type, type);
   }
 
+  public boolean isErrorProneImmutable() {
+    return passedAnnotations().contains("@com.google.errorprone.annotations.Immutable");
```

Comment:
This `#contains` check is the reason for the `CharSequence` -> `String` changes. In practice the `CharSequence` instances were `StringBuilder`s, and those don't implement `#equals`. If you'd like me to implement this some other way, just let me know :)

Reply:
Thank you, overall, I'm ok with the idea. (Sorry for not getting to this earlier)
Looking at this, just to mention, `CharSequence/StringBuilder` was to try to avoid tons of strings copying, but actual effect/impact have not investigated. As for me it seems that it's a little strange to piggy-back on `passedAnnotations()` accessor as it's complicating things even more (fully realizing that a lot of things are already complicated ;) ). Seems that dedicated support for detecting `com.google.errorprone.annotations.Immutable` and copying it over would be little bit more straightforward. You can just hook into `ValueType` and detect it's presence on `element` via lazy getter, or/and for example, ask it's present in `constitution.style().passAnnotationsNames()` (but may not look into this property at all), or auto-detect on the classpath and apply if necessary (using accessor to private `Environment.findElement`). I'm a bit confused if we would auto-add this to implementation or add it only if it's placed on abstract value type, but it can work without changing CharSequences to Strings (not because it was a good idea to use CharSequences, but just because it's unnecessary twist here, IMHO)

Reply:
Thanks for the feedback! I'll have to open the code in my IDE to cross-reference your suggestions; hope to get back to this later this week :)

Reply:
This PR slipped my attention; it's been very busy at work recently. Commit google/error-prone@e7daed7e102c341230905cc07872b778250c4193 was mirrored out yesterday; that makes this PR quite a bit more pressing... so will try to find time to finalize it.